### PR TITLE
[ADVAPP-710]: Allow a Basic Needs Program to be associated to a Student

### PR DESCRIPTION
**Pull request by payal-canyon**
### Ticket(s) or GitHub Issue

- https://canyongbs.atlassian.net/browse/ADVAPP-710

### Technical Description

> Describe your changes in detail. This should include technical/architectural decisions and reasoning, if applicable.
> Programs can be linked to student profiles, providing advisors with a quick view of the resources available to individual students based on their needs.

_______________________________________________

### CODE REVIEW

File: app-modules/basic-needs/database/migrations/2024_07_02_161929_create_basic_needs_program_student_table.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+*/
+
+use Illuminate\Support\Facades\Schema;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Database\Migrations\Migration;
+
+return new class () extends Migration {
+    public function up(): void
+    {
+        Schema::create('basic_needs_program_student', function (Blueprint $table) {
```
[the remark this answers is not in the document]
Comment (Orrison) [reply app-modules/basic-needs/database/migrations/2024_07_02_161929_create_basic_needs_program_student_table.php:44]: @payal-canyon After consulting with Product, we will eventually have Programs related to Prospects. So in order to prepare for that now let's change this Pivot table and the models to support a constrained polymorphic many-to-many relationship: https://laravel.com/docs/11.x/eloquent-relationships#many-to-many-polymorphic-relations

The `-able` naming convention Laravel often proposes doesn't really make sense in this context though. So you may consider calling the pivot table `program_participants`. Then the relationship for Student "programs" and from Programs call the relationship "participants"

File: app-modules/basic-needs/src/Filament/Resources/BasicNeedsProgramResource/Pages/ManageStudents.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+{
+    protected static string $resource = BasicNeedsProgramResource::class;
+
+    protected static string $relationship = 'students';
+
+    protected static ?string $navigationIcon = 'heroicon-o-academic-cap';
+
+    public static function getNavigationLabel(): string
+    {
+        return 'Students';
```
Comment (Orrison) [inline app-modules/basic-needs/src/Filament/Resources/BasicNeedsProgramResource/Pages/ManageStudents.php:54]: Let's call this `Participants`

File: app-modules/basic-needs/src/Filament/Resources/BasicNeedsProgramResource/Pages/ManageStudents.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+use AdvisingApp\BasicNeeds\Filament\Resources\BasicNeedsProgramResource;
+use AdvisingApp\BasicNeeds\Filament\Resources\BasicNeedsProgramResource\RelationManagers\StudentsRelationManager;
+
+class ManageStudents extends ManageRelatedRecords
+{
+    protected static string $resource = BasicNeedsProgramResource::class;
+
+    protected static string $relationship = 'students';
+
+    protected static ?string $navigationIcon = 'heroicon-o-academic-cap';
```
Comment (Orrison) [inline app-modules/basic-needs/src/Filament/Resources/BasicNeedsProgramResource/Pages/ManageStudents.php:50]: I think you need to add something like `protected static ?string $breadcrumb = 'Participants';` here to get better breadcrumbs

File: app-modules/student-data-model/src/Filament/Resources/StudentResource/Pages/ManagePrograms.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+use AdvisingApp\StudentDataModel\Filament\Resources\StudentResource;
+use AdvisingApp\StudentDataModel\Filament\Resources\BasicNeedsProgramResource\RelationManagers\BasicNeedsProgramsRelationManager;
+
+class ManagePrograms extends ManageRelatedRecords
+{
+    protected static string $resource = StudentResource::class;
+
+    protected static string $relationship = 'basicNeedsPrograms';
+
+    protected static ?string $navigationIcon = 'heroicon-o-square-3-stack-3d';
```
Comment (Orrison) [inline app-modules/student-data-model/src/Filament/Resources/StudentResource/Pages/ManagePrograms.php:50]: I think you may need to add something like `protected static ?string $breadcrumb = 'Programs';` here to get better bread crumbs

**Review by Orrison — CHANGES_REQUESTED** [review]
Left some notes. Please also review the most recent comment from Product in the ticket as they need some changes to the first screenshot once my suggested changes in the PR are done.

File: app-modules/basic-needs/src/Filament/Resources/BasicNeedsProgramResource/Pages/ManageStudents.php (hunk 3 above)
Comment (payal-canyon) [reply app-modules/basic-needs/src/Filament/Resources/BasicNeedsProgramResource/Pages/ManageStudents.php:50]: This file is now removed due to relationship changes.

File: app-modules/student-data-model/src/Filament/Resources/StudentResource/Pages/ManagePrograms.php (hunk 4 above)
Comment (payal-canyon) [reply app-modules/student-data-model/src/Filament/Resources/StudentResource/Pages/ManagePrograms.php:50]: This file is now removed due to relationship changes.

File: app-modules/basic-needs/database/migrations/2024_07_05_072607_create_program_participants_table.php [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Database\Migrations\Migration;
+
+return new class () extends Migration {
+    public function up(): void
+    {
+        Schema::create('program_participants', function (Blueprint $table) {
+            $table->foreignUuid('basic_needs_program_id')->constrained('basic_needs_programs')->cascadeOnDelete();
+            $table->string('program_participants_type');
+            $table->string('program_participants_id');
```
Comment (Orrison) [inline app-modules/basic-needs/database/migrations/2024_07_05_072607_create_program_participants_table.php:47]: Let's add an index for these like `$table->index(['program_participants_type', 'program_participants_id']);` at the bottom of the function.

File: app-modules/prospect/src/Filament/Resources/ProspectResource/Pages/ManageProspectPrograms.php [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+    public function getRelationManagers(): array
+    {
+        return [ProgramRelationManager::class];
+    }
+
+    public static function canAccess(array $parameters = []): bool
+    {
+        if (Feature::active('manage-program-participants')) {
+            return true;
```
Comment (Orrison) [inline app-modules/prospect/src/Filament/Resources/ProspectResource/Pages/ManageProspectPrograms.php:69]: This should return the parent `canAccess`

File: app-modules/basic-needs/src/Filament/Resources/BasicNeedsProgramResource/Pages/ManageParticipants.php [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+            return $cluster::unshiftClusterBreadcrumbs($breadcrumbs);
+        }
+
+        return $breadcrumbs;
+    }
+
+    public static function canAccess(array $parameters = []): bool
+    {
+        if (Feature::active('manage-program-participants')) {
+            return true;
```
Comment (Orrison) [inline app-modules/basic-needs/src/Filament/Resources/BasicNeedsProgramResource/Pages/ManageParticipants.php:78]: This should return the parent `canAccess`

File: app-modules/student-data-model/src/Filament/Resources/StudentResource/Pages/ManageStudentPrograms.php [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+    public function getRelationManagers(): array
+    {
+        return [ProgramRelationManager::class];
+    }
+
+    public static function canAccess(array $parameters = []): bool
+    {
+        if (Feature::active('manage-program-participants')) {
+            return true;
```
Comment (Orrison) [inline app-modules/student-data-model/src/Filament/Resources/StudentResource/Pages/ManageStudentPrograms.php:69]: This should return the parent `canAccess`

**Review by Orrison — CHANGES_REQUESTED** [review]

File: app-modules/student-data-model/src/Filament/Resources/StudentResource/Pages/ManageStudentPrograms.php (hunk 8 above)
Comment (payal-canyon) [reply app-modules/student-data-model/src/Filament/Resources/StudentResource/Pages/ManageStudentPrograms.php]: Hi @Orrison ,
Directly parent::canAccess() is giving forbidden errors. So, I passed the $parameters into that to fix the forbidden issue.

Thank You!

**Review by Orrison — APPROVED** [review]

**Pull request merged**